### PR TITLE
Fix binlookup

### PR DESCRIPTION
**Pull request by descorp**
### Changes

* fix payload encryption on BinLookup
* add throttling to BinLookup request

### CODE REVIEW

File: AdyenCard/Components/Card/CardComponent.swift [hunk 1]
```diff
@@ -4,7 +4,7 @@
 // This file is open source and available under the MIT license. See the LICENSE file for more info.
 //
 
-import Foundation
+import Adyen
```
Comment (mohammedDehairy) [inline AdyenCard/Components/Card/CardComponent.swift:7]: why is this changed?

File: AdyenCard/Components/Card/CardComponent.swift [hunk 2]
```diff
@@ -220,10 +221,7 @@ public final class CardComponent: PaymentComponent, PresentableComponent, Locali
                                       localizationParameters: localizationParameters)
         
         observe(item.$binValue) { [weak self] bin in
-            guard let self = self else { return }
-            
-            self.requestCardType(for: bin, update: item)
-            self.cardComponentDelegate?.didChangeBIN(String(bin.prefix(CardComponent.publicBinLenght)), component: self)
+            self?.binDidReceived(bin)
```
Comment (mohammedDehairy) [inline AdyenCard/Components/Card/CardComponent.swift:224]: `self?.didReceive(bin)` sounds much better IMO :-)

File: AdyenCard/Components/Card/CardComponent.swift [hunk 3]
```diff
@@ -282,12 +280,22 @@ public final class CardComponent: PaymentComponent, PresentableComponent, Locali
...
+    
+    private func requestCardTypes(for bin: String) {
         cardTypeProvider.requestCardType(for: bin, supported: self.supportedCardTypes) { [weak self] cardTypes in
             guard let self = self else { return }
             
-            self.securityCodeItem.selectedCard = cardTypes.first
-            item.detectedCardsDidChange(detectedCards: cardTypes)
+            self.numberItem.detectedCardsDidChange(detectedCards: cardTypes)
             self.cardComponentDelegate?.didChangeCardType(cardTypes, component: self)
         }
```
Comment (mohammedDehairy) [inline AdyenCard/Components/Card/CardComponent.swift:300]: what happens if there is no internet connection, or if the call failed for any reason, should we report the locally detected card types?

File: Adyen/Utilities/Throttler.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
+//
+// Copyright (c) 2020 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+// Throttles requests.
+public final class Throttler {
```
Comment (mohammedDehairy) [inline Adyen/Utilities/Throttler.swift:10]: why do we need to throttle bin lookup calls?

File: AdyenCard/Components/Card/CardComponent.swift (hunk 3 above)
Comment (descorp) [reply AdyenCard/Components/Card/CardComponent.swift:300]: it will do so. there are two fallback inside `cardTypeProvider ` : on "no publicKey" and "unsuccessful binLookup request"

File: Adyen/Utilities/Throttler.swift (hunk 4 above)
Comment (descorp) [reply Adyen/Utilities/Throttler.swift:10]: you type 16 digits + make mistakes and corrections. Every change is a network call.
Throttling will save you some network data.

Comment (descorp) [reply Adyen/Utilities/Throttler.swift:10]: Alternatively, we can move it into CardNumberItem 🤔

File: AdyenCard/Components/Card/CardComponent.swift (hunk 1 above)
Comment (descorp) [reply AdyenCard/Components/Card/CardComponent.swift:7]: good question :D

File: AdyenCard/Components/Card/CardComponent.swift (hunk 2 above)
Comment (descorp) [reply AdyenCard/Components/Card/CardComponent.swift:224]: 👌

File: Adyen/Utilities/Throttler.swift (hunk 4 above)
Comment (mohammedDehairy) [reply Adyen/Utilities/Throttler.swift:10]: hmm 🤔 not sure if its worth the complexity TBH, it will save the shopper maximum 100KB of data? is this worth a delay 0.5 between calls and the complexity? this is not a rhetorical question 😊.

Comment (mohammedDehairy) [reply Adyen/Utilities/Throttler.swift:10]: ok, I guess it make sense :-)

File: AdyenCard/Components/Card/CardComponent.swift [hunk 5]
```diff
@@ -282,12 +280,22 @@ public final class CardComponent: PaymentComponent, PresentableComponent, Locali
...
+    }
+    
+    private func requestCardTypes(for bin: String) {
         cardTypeProvider.requestCardType(for: bin, supported: self.supportedCardTypes) { [weak self] cardTypes in
             guard let self = self else { return }
             
-            self.securityCodeItem.selectedCard = cardTypes.first
-            item.detectedCardsDidChange(detectedCards: cardTypes)
+            self.numberItem.detectedCardsDidChange(detectedCards: cardTypes)
             self.cardComponentDelegate?.didChangeCardType(cardTypes, component: self)
```
Comment (mohammedDehairy) [inline AdyenCard/Components/Card/CardComponent.swift:299]:
```
self.numberItem.didChange(detectedCards: cardTypes)
```
would be more swifty IMO.

also 

```
self.cardComponentDelegate?.didChange(cardTypes, component: self)
```

but I guess its too late for this one, right?

File: Adyen/Utilities/Throttler.swift (hunk 4 above)
Comment (mohammedDehairy) [inline Adyen/Utilities/Throttler.swift:10]: does this need to be `public`?

Comment (descorp) [reply Adyen/Utilities/Throttler.swift:10]: nope, it only used in AdyenCard

File: AdyenCard/Form/FormCardNumberItem.swift [hunk 6]
```diff
@@ -49,21 +49,20 @@ internal final class FormCardNumberItem: FormTextItem {
...
         }
     }
     
-    internal func detectedCardsDidChange(detectedCards: [CardType]) {
+    internal func didChange(detectedCards: [CardType]) {
         let newDetectedCardTypes = Set<CardType>(detectedCards)
         guard self.previouslyDetectedCardTypes != newDetectedCardTypes else { return }
         
         previouslyDetectedCardTypes = newDetectedCardTypes
-        cardNumberFormatter.cardType = newDetectedCardTypes.first
```
Comment (mohammedDehairy) [inline AdyenCard/Form/FormCardNumberItem.swift:66]: why remove this line? the formatter should adjust to the true card types detected by the backend, right?

File: AdyenCard/Utilities/CardType/CardTypeProvider.swift [hunk 7]
```diff
@@ -32,45 +38,50 @@ internal final class CardTypeProvider: Component {
     /// - Parameters:
     ///   - bin: Card's BIN number. If longer than `minBinLength` - calls API, otherwise check local Regex,
     ///   - caller:  Callback to notify about results.
```
Comment (mohammedDehairy) [inline AdyenCard/Utilities/CardType/CardTypeProvider.swift:40]: maybe change the docs as well: `caller -> completion`

Comment (descorp) [reply AdyenCard/Utilities/CardType/CardTypeProvider.swift:40]: It moved to line 52. `didChange (detectedCards:)` is not looking nicely under throttling, so I've moved it out.

File: AdyenCard/Components/Card/CardComponent.swift (hunk 5 above)
Comment (descorp) [reply AdyenCard/Components/Card/CardComponent.swift:299]: renamed

File: AdyenTests/Card Tests/Mocks/PublicKeyProviderMock.swift [hunk 8]
```diff
@@ -1,9 +1,7 @@
...
-//  PublicKeyProviderMock.swift
-//  AdyenTests
+// Copyright (c) 2020 Adyen N.V.
 //
-//  Created by Vladimir Abramichev on 24/08/2020.
-//  Copyright © 2020 Adyen. All rights reserved.
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
 //
 
 @testable import Adyen
```
Comment (mohammedDehairy) [inline AdyenTests/Card Tests/Mocks/PublicKeyProviderMock.swift:7]: do we still need this :-) ?

File: AdyenTests/Card Tests/Utilities/CardEncryptorCardTests.swift [hunk 9]
```diff
@@ -145,4 +145,36 @@ class CardEncryptorCardTests: XCTestCase {
 
         XCTAssertNotNil( try? CardEncryptor.encryptedToken(for: card, holderName: nil, publicKey: key))
     }
+
+    // MARK: - Test encrypting BIN
+
+    func testEncryptExpiryBINShouldReturnNilWithNilBIN() {
```
Comment (mohammedDehairy) [inline AdyenTests/Card Tests/Utilities/CardEncryptorCardTests.swift:151]: I think the name should be `testEncryptBINShouldReturnNilWithEmptyBIN`, right?

File: AdyenCard/Form/FormCardNumberItem.swift (hunk 6 above)
Comment (descorp) [reply AdyenCard/Form/FormCardNumberItem.swift:66]: It moved to line 52. `cardNumberFormatter` is mostly detecting Amex / no-Amex, so local regex-based recognition is more appropriate.

File: AdyenTests/Card Tests/Utilities/CardTypeProviderTests.swift [hunk 10]
```diff
@@ -1,9 +1,7 @@
...
-//  CardTypeProviderTests.swift
-//  AdyenTests
+// Copyright (c) 2020 Adyen N.V.
 //
-//  Created by Vladimir Abramichev on 24/08/2020.
-//  Copyright © 2020 Adyen. All rights reserved.
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
 //
 
 @testable import AdyenCard
```
Comment (mohammedDehairy) [inline AdyenTests/Card Tests/Utilities/CardTypeProviderTests.swift:7]: empty file :-)

File: AdyenCard/Utilities/CardType/CardTypeProvider.swift (hunk 7 above)
Comment (descorp) [reply AdyenCard/Utilities/CardType/CardTypeProvider.swift:40]: ups :D

File: AdyenTests/Card Tests/Mocks/PublicKeyProviderMock.swift (hunk 8 above)
Comment (descorp) [reply AdyenTests/Card Tests/Mocks/PublicKeyProviderMock.swift:7]: copy paste..

File: AdyenTests/Card Tests/Utilities/CardTypeProviderTests.swift (hunk 10 above)
Comment (descorp) [reply AdyenTests/Card Tests/Utilities/CardTypeProviderTests.swift:7]: what do you mean?

File: AdyenTests/Card Tests/Utilities/CardEncryptorCardTests.swift (hunk 9 above)
Comment (descorp) [reply AdyenTests/Card Tests/Utilities/CardEncryptorCardTests.swift:151]: why not

File: AdyenCard/Form/FormCardNumberItem.swift (hunk 6 above)
Comment (mohammedDehairy) [reply AdyenCard/Form/FormCardNumberItem.swift:66]: 👍 ok

File: AdyenTests/Card Tests/Utilities/ThrottlerTests.swift [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+            }
+        }
+
+        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(1)) {
+            expectation.fulfill()
+        }
+
+        wait(for: [expectation], timeout: 2)
+        XCTAssertEqual(counter, 1)
+    }
```
Comment (mohammedDehairy) [inline AdyenTests/Card Tests/Utilities/ThrottlerTests.swift:30]: nice test 👍

File: AdyenTests/Card Tests/Utilities/CardTypeProviderTests.swift (hunk 10 above)
Comment (mohammedDehairy) [reply AdyenTests/Card Tests/Utilities/CardTypeProviderTests.swift:7]: sorry, I mean, why change the file header comments?

Comment (descorp) [reply AdyenTests/Card Tests/Utilities/CardTypeProviderTests.swift:7]: it is nicer :D

**Review by mohammedDehairy — APPROVED** [review]
